src/NHibernate.Test/Linq/QueryLock.cs

```diff
@@ -0,0 +1,235 @@
+					.SetLockMode("o", LockMode.Upgrade)
+					.List();
+			}
+		}
+
+		[Test]
+		public void CanSetLockOnJoin()
+		{
+			using (session.BeginTransaction())
+			{
+				var result = (from c in db.Customers
+				              from o in c.Orders.WithLock(LockMode.Upgrade)
+				              select o).ToList();
+
+				Assert.That(result, Has.Count.EqualTo(830));
+				Assert.That(session.GetCurrentLockMode(result[0]), Is.EqualTo(LockMode.Upgrade));
+			}
+		}
+
+		[Test]
+		public void CanSetLockOnJoinOuter()
+		{
+			using (session.BeginTransaction())
+			{
+				var result = (from c in db.Customers
+				              from o in c.Orders
+				              select o).WithLock(LockMode.Upgrade).ToList();
+
+				Assert.That(result, Has.Count.EqualTo(830));
+				Assert.That(session.GetCurrentLockMode(result[0]), Is.EqualTo(LockMode.Upgrade));
+			}
+		}
+
+		[Test]
+		public void CanSetLockOnJoinOuterNotSupported()
+		{
+			using (session.BeginTransaction())
+			{
+				var query = (
+					from c in db.Customers
+					from o in c.Orders
+					select new {o, c}
+				).WithLock(LockMode.Upgrade);
+
+				Assert.Throws<NotSupportedException>(() => query.ToList());
+			}
+		}
+
+		[Test]
+		public void CanSetLockOnJoinOuter2Hql()
+		{
+			using (session.BeginTransaction())
+			{
+				session
+					.CreateQuery("select o, c from Customer c join c.Orders o")
+					.SetLockMode("o", LockMode.Upgrade)
+					.SetLockMode("c", LockMode.Upgrade)
+					.List();
+			}
+		}
+
+		[Test]
+		public void CanSetLockOnBothJoinAndMain()
+		{
+			using (session.BeginTransaction())
+			{
+				var result = (
+					from c in db.Customers.WithLock(LockMode.Upgrade)
+					from o in c.Orders.WithLock(LockMode.Upgrade)
+					select new {o, c}
+				).ToList();
+
+				Assert.That(result, Has.Count.EqualTo(830));
+				Assert.That(session.GetCurrentLockMode(result[0].o), Is.EqualTo(LockMode.Upgrade));
+				Assert.That(session.GetCurrentLockMode(result[0].c), Is.EqualTo(LockMode.Upgrade));
+			}
+		}
+
+		[Test]
+		public void CanSetLockOnBothJoinAndMainComplex()
+		{
+			using (session.BeginTransaction())
+			{
+				var result = (
+					from c in db.Customers.Where(x => true).WithLock(LockMode.Upgrade)
+					from o in c.Orders.Where(x => true).WithLock(LockMode.Upgrade)
+					select new {o, c}
+				).ToList();
+
+				Assert.That(result, Has.Count.EqualTo(830));
+				Assert.That(session.GetCurrentLockMode(result[0].o), Is.EqualTo(LockMode.Upgrade));
+				Assert.That(session.GetCurrentLockMode(result[0].c), Is.EqualTo(LockMode.Upgrade));
+			}
+		}
+
+		[Test]
+		public void CanSetLockOnLinqPagingQuery()
+		{
+			using (session.BeginTransaction())
+			{
+				var result = (from e in db.Customers
+				              select e).Skip(5).Take(5).WithLock(LockMode.Upgrade).ToList();
+
+				Assert.That(result, Has.Count.EqualTo(5));
+				Assert.That(session.GetCurrentLockMode(result[0]), Is.EqualTo(LockMode.Upgrade));
+
+				if (!(Dialect is MsSql2000Dialect) && !(Dialect is MsSqlCeDialect))
```

Comment:
This would require 2 separate test dialect properties because those dialects do support locks in general, but they do not  support them in some limited circumstances (n subqueries, for example).